### DISCUSSION TITLE
efficient computation of block diagonal matrix vector multiplication when matrix is in compressed format

### DISCUSSION THREAD

**Discussion by parshakova** [jax-ml/jax#14288] [Q&A]
Given matrix $B \in \mathrm{R}^{n \times r}$, a vector $x \in \mathrm{R}^{n}$ and a set of indices for slicing matrix $i_1, \ldots,i_k$. 

We want to compute the following matvec efficiently using jax 
`block_diag`$(B_{:i_1}, \ldots, B_{i_{k-1}:i_k})^T x$,
where $B$ is being sliced along the first dimension and slices are **not** of the same size.

I tried these two functions but it is pretty slow

```python
n = 5000
r = 50
x = jnp.asarray(np.random.randn(n,1))
B = jnp.asarray(np.random.randn(n,r))
indices = [1200, 2500, 2800, 4000]

def block_diag_expand(B, x, indices):
    blocks = jnp.split(B, indices, axis=0)
    res = jnp.dot(x.T, jax.scipy.linalg.block_diag(*blocks))
    return res.T

def matvec_slicing(B, x, indices):
    vec = np.zeros(x.shape)
    B_blocks = jnp.split(B, indices, axis=0)
    x_blocks = jnp.split(x, indices, axis=0)
    res = []
    for x_b, B_b in zip(x_blocks, B_blocks):
        res += [jnp.dot(x_b.T, B_b).T]
    return jnp.concatenate(res, axis=0)

%timeit block_diag_expand(B, x, indices)
# 1.53 ms ± 76.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

%timeit matvec_slicing(B, x, indices)
# 3.72 ms ± 416 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
the process above takes $n r$ flops, therefore those functions are super slow, e.g., comparing to $n r$ flops 
```
%timeit jnp.dot(x.T, B).T
353 µs ± 43.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Does anyone have any idea on how to improve this?
I did not figure out how to use `vmap` in this case when the slices are of different sizes, because we cannot just unfold the vector and matrix and do some kind of batch matrix vector multiplication.

**Comment by jakevdp** [comment]
The variable slice sizes make this somewhat more difficult, but often you can use `jax.ops.segment_sum` or related operations in situations like this. Here's one possible approach
```python
@jax.jit
def segment_sum(B, x, indices):
  i = jnp.zeros(B.shape[0], int).at[jnp.array(indices)].set(1).cumsum()
  return jax.ops.segment_sum(B * x, i, num_segments=1 + len(indices)).reshape(-1, 1)

ans1 = block_diag_expand(B, x, indices)
ans2 = matvec_slicing(B, x, indices)
ans3 = segment_sum(B, x, indices)

np.testing.assert_allclose(ans1, ans2)
np.testing.assert_allclose(ans1, ans3, rtol=1E-4)  # larger tolerance due to float accumulation errors

%timeit block_diag_expand(B, x, indices).block_until_ready()
# 4.7 ms ± 947 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit matvec_slicing(B, x, indices).block_until_ready()
# 2.77 ms ± 114 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit segment_sum(B, x, indices).block_until_ready()
# 640 µs ± 78.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Reply by parshakova** [reply]
Thank you very much!

Would you have a suggestion for fast implementation of the related problem below?
 Namely, we are given matrices  $B \in \mathrm{R}^{m\times r}$ and $C \in \mathrm{R}^{n\times r}$ and indices for slicing the columns $i_1, \ldots, i_L$ such that $C_{:,i_l:i_{l+1}}  \in \mathrm{R}^{n\times r_l}$ and $B_{:,i_l:i_{l+1}}  \in \mathrm{R}^{m\times r_l}$ with  $r_l=i_{l+1}-i_l$ and $\sum_{l}r_l = r$. Define block matrices $C^l =C_{:,i_l:i_{l+1}}$ and $B^l =B_{:,i_l:i_{l+1}}$, then $B = [B^1, \ldots, B^L]$ and $C = [C^1, \ldots, C^L]$.
As before we have a vector $x \in \mathrm{R}^{n}$ and a set of indices for each $B_l$ for slicing the rows $b^l_1, \ldots, b^l_k$ and similarly indices $c_1, \ldots, c_L$ for slicing rows of $C^l$.
We want to compute the following operation efficiently using jax
$$\sum_{l}  \textrm{blockdiag}(B^l_{:b^l_1}, \ldots, B^l_{b^l_{k-1}:b^l_k}) \textrm{blockdiag}(C^l_{:c^l_1}, \ldots, C^l_{c^l_{k-1}:c^l_k})^T x,$$
where the slices are **not** of the same size.

```python
from jax.config import config
config.update("jax_enable_x64", True)

n = 5000
m = 4000
r = 50
x = jnp.asarray(np.random.randn(n, 1),dtype=jnp.float64)
B = jnp.asarray(np.random.randn(m, r),dtype=jnp.float64)
C = jnp.asarray(np.random.randn(n, r),dtype=jnp.float64)

indices_r = [0, r//7, r//2, 2*r//3, r]
indices_n = [[0, n], [0, n//2, n], [0, n//7 ,n//3, 2*n//3, n], [0, n//7 ,n//3, 2*n//3, n]]
indices_m = [[0, m], [0, m//3, m], [0, m//6, m//3, 2*m//3, m], [0, m//6, m//3, 2*m//3, m]]

def block_diag_expand(B, x, indices, reverse=False):
    blocks = jnp.split(B, indices[1:-1], axis=0)
    if reverse:
        res = jnp.dot(jax.scipy.linalg.block_diag(*blocks), x)
    else:
        res = jnp.dot(x.T, jax.scipy.linalg.block_diag(*blocks)).T
    return res

def matvec_slicing(B, x, indices, reverse=False):
    vec = np.zeros(x.shape); res = []
    B_blocks = jnp.split(B, indices[1:-1], axis=0)
    if reverse:
        x_blocks = jnp.split(x, jnp.arange(0, x.size, B.shape[1])[1:], axis=0)
    else:
        x_blocks = jnp.split(x, indices[1:-1], axis=0)
    for x_b, B_b in zip(x_blocks, B_blocks):
        if reverse: res += [jnp.dot(B_b, x_b)]
        else: res += [jnp.dot(x_b.T, B_b).T]
    return jnp.concatenate(res, axis=0)

@jax.jit
def segment_sum0(B, x, indices):
    i = jnp.zeros(B.shape[0], int).at[jnp.array(indices[1:])].set(1).cumsum()
    res = jax.ops.segment_sum(jnp.multiply(B,x), i, num_segments=len(indices)-1).reshape(-1, 1)
    return res

# @jax.jit
def segment_sum(B, x, indices, reverse=False):
    if reverse:
        dn = ('NHC', 'OIH', 'NHC')
        lhs = x.reshape((1, -1, 1))
        patches = jax.lax.conv_general_dilated_patches(
            lhs=lhs, filter_shape=(B.shape[1],), window_strides=(B.shape[1],),
            padding='VALID', dimension_numbers=dn)[0]
        rep_x = jnp.repeat(patches, repeats=np.diff(np.array(indices)), axis=0)
        res = jnp.multiply(B,rep_x).sum(axis=1, keepdims=True)
        return res
    else:
        return segment_sum0(B, x, indices)

def nested_BCx(func, C, B, x, indices_r, indices_m, indices_n):
    res = 0
    for i in range(len(indices_r)-1):
        C_lx = func(C[:,indices_r[i]:indices_r[i+1]], x, indices_n[i])
        res += func(B[:,indices_r[i]:indices_r[i+1]], C_lx, indices_m[i], reverse=True)
    return res
    
ans1 = nested_BCx(block_diag_expand, C, B, x, indices_r, indices_m, indices_n)
ans2 = nested_BCx(matvec_slicing, C, B, x, indices_r, indices_m, indices_n)
ans3 = nested_BCx(segment_sum, C, B, x, indices_r, indices_m, indices_n)

print(jnp.linalg.norm(ans1-ans2), jnp.linalg.norm(ans1-ans3))
%timeit nested_BCx(block_diag_expand, C, B, x, indices_r, indices_m, indices_n)
# 16.3 ms ± 860 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit nested_BCx(matvec_slicing, C, B, x, indices_r, indices_m, indices_n)
# 28.1 ms ± 3.23 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit nested_BCx(segment_sum, C, B, x, indices_r, indices_m, indices_n)
# 45.3 ms ± 3.15 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Do you have any recommendations on how to keep good accuracy without the double precision?
The difference in answers when using `float32` is `0.0015` which is not at all negligible.
